File: app/src/debug/java/sk/vander/cmind/data/api/DebugApiModule.java
package sk.vander.cmind.data.api;

import com.facebook.stetho.okhttp3.StethoInterceptor;

import dagger.Module;
import dagger.Provides;
import okhttp3.Cache;
import okhttp3.Interceptor;
import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.mock.MockRetrofit;
import retrofit2.mock.NetworkBehavior;
import sk.vander.library.annotation.ApplicationScope;
import sk.vander.library.prefs.BooleanPreference;

/**
 * Created by arashid on 27/06/16.
 */
@Module(includes = ApiModule.class)
public class DebugApiModule {

  @Provides @ApplicationScope @MockMode Boolean provideMockMode(@MockMode BooleanPreference mockPref) {
    return mockPref.get();
  }

  @Provides @ApplicationScope HttpLoggingInterceptor.Level provideLoggingLevel() {
    return HttpLoggingInterceptor.Level.BODY;
  }

  @ApplicationScope @Provides HttpLoggingInterceptor provideLoggingInterceptor(HttpLoggingInterceptor.Level level) {
    final HttpLoggingInterceptor interceptor = new HttpLoggingInterceptor();
    interceptor.setLevel(level);
    return interceptor;
  }

  @ApplicationScope @Provides OkHttpClient provideOkHttpClient(Cache cache,
                                                               HttpLoggingInterceptor loggingInterceptor,
                                                               @ForceCacheInterceptor Interceptor forceCache) {
    return new OkHttpClient.Builder()
        .addNetworkInterceptor(new StethoInterceptor())
//        .addNetworkInterceptor(forceCache)
        .addInterceptor(loggingInterceptor)
        .cache(cache)
        .build();
  }

  @ApplicationScope @Provides NetworkBehavior provideNetworkBehavior() {
    return NetworkBehavior.create();
  }

  @ApplicationScope @Provides MockRetrofit provideMockRetrofit(Retrofit retrofit, NetworkBehavior networkBehavior) {
    return new MockRetrofit.Builder(retrofit).networkBehavior(networkBehavior).build();
  }
}

File: app/src/debug/java/sk/vander/cmind/data/api/DebugApiModule.kt
package sk.vander.cmind.data.api

import com.facebook.stetho.okhttp3.StethoInterceptor

import dagger.Module
import dagger.Provides
import okhttp3.Cache
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.mock.MockRetrofit
import retrofit2.mock.NetworkBehavior
import sk.vander.library.annotation.ApplicationScope
import sk.vander.library.prefs.BooleanPreference

/**
 * Created by arashid on 27/06/16.
 */
@Module(includes = arrayOf(ApiModule::class))
class DebugApiModule {

  @Provides @ApplicationScope @MockMode fun provideMockMode(@MockMode mockPref: BooleanPreference): Boolean? {
    return mockPref.get()
  }

  @Provides @ApplicationScope fun provideLoggingLevel(): HttpLoggingInterceptor.Level {
    return HttpLoggingInterceptor.Level.BODY
  }

  @ApplicationScope @Provides fun provideLoggingInterceptor(level: HttpLoggingInterceptor.Level): HttpLoggingInterceptor {
    val interceptor = HttpLoggingInterceptor()
    interceptor.level = level
    return interceptor
  }

  @ApplicationScope @Provides fun provideOkHttpClient(cache: Cache,
                                                               loggingInterceptor: HttpLoggingInterceptor,
                                                               @ForceCacheInterceptor forceCache: Interceptor): OkHttpClient {
    return OkHttpClient.Builder().addNetworkInterceptor(StethoInterceptor()).addInterceptor(loggingInterceptor)//        .addNetworkInterceptor(forceCache)
        .cache(cache).build()
  }

  @ApplicationScope @Provides fun provideNetworkBehavior(): NetworkBehavior {
    return NetworkBehavior.create()
  }

  @ApplicationScope @Provides fun provideMockRetrofit(retrofit: Retrofit, networkBehavior: NetworkBehavior): MockRetrofit {
    return MockRetrofit.Builder(retrofit).networkBehavior(networkBehavior).build()
  }
}
